Tests/LibAudio/TestWav.cpp

```diff
@@ -0,0 +1,167 @@
+    #ifdef AK_OS_SERENITY
+        ByteString path = ByteString::formatted("/usr/Tests/LibAudio/wav-test-files/{}", file_name);
+        ByteString out_path = ByteString::formatted("/home/anon/saved_{}", file_name);
+    #else
+        ByteString path = ByteString::formatted("wav-test-files/{}", file_name);
+        ByteString out_path = ByteString::formatted("saved_{}", file_name);
+    #endif
+
+    auto maybe_loader = Audio::Loader::create(path);
+    if (maybe_loader.is_error()) {
+        FAIL(maybe_loader.error());
+        return;
+    }
+    auto loader = maybe_loader.value();
+
+    EXPECT_EQ(loader->format_name(), format);
+    EXPECT_EQ(loader->sample_rate(), rate);
+    EXPECT_EQ(loader->num_channels(), channels);
+    EXPECT_EQ(loader->bits_per_sample(), bits);
+    EXPECT_EQ(loader->total_samples(), num_samples);
+
+    auto maybe_writer = Audio::WavWriter::create_from_file(out_path, rate, channels);
+    if (maybe_writer.is_error()) {
+        FAIL(maybe_writer.error());
+        return;
+    }
+    auto writer = maybe_writer.release_value();
+
+    u32 sum=0;
+    int samples_read=0;
+    int size=0;
+    do {
+        auto maybe_samples = loader->get_more_samples(1000);
+        auto samples = maybe_samples.release_value();
+
+        ErrorOr<void> maybe_error = writer->write_samples(samples.span());
+        if (maybe_error.is_error()) {
+            FAIL(maybe_error.error());
+        }
+        size = samples.size();
+
+        if (size) {
+            samples_read += size;
+            for (int i=0;i<size;i++) {
+                sum += samples[i].left * 32767 + 32768;
+                sum += samples[i].right * 32767 + 32768;
+            }
+        }
+
+    } while (size);
+
+    ErrorOr<void> maybe_error = writer->finalize();
+    if (maybe_error.is_error()) {
+        FAIL("Could not finalise writer");
+    }
+
+    EXPECT_EQ(samples_read, num_samples);
+    EXPECT_EQ(sum, checksum);
+
+    maybe_loader = Audio::Loader::create(out_path);
+    if (maybe_loader.is_error()) {
+        FAIL(maybe_loader.error());
+        return;
+    }
+    loader = maybe_loader.value();
+
+    u32 new_sum=0;
+    int new_samples_read=0;
+    int new_size=0;
+
+    FixedArray<Audio::Sample> samples;
+    do {
+        auto maybe_samples = loader->get_more_samples(1000);
+        if (maybe_samples.is_error()) {
+            FAIL(maybe_samples.release_error());
+            return;
+        }
+        auto samples = maybe_samples.release_value();
+
+        new_size = samples.size();
+        if (new_size) {
+            new_samples_read += new_size;
+            for (int i=0;i<new_size;i++) {
+                new_sum += samples[i].left * 32767 + 32768;
+                new_sum += samples[i].right * 32767 + 32768;
+            }
+        }
+    } while (new_size);
+
+    EXPECT_EQ(new_samples_read, num_samples);
+    EXPECT_EQ(new_sum, checksum);
+
+    unlink(out_path.characters());
```

Comment:
This will only happen if the test succeeds and nothing crashes beforehand. Please use a ScopeGuard at the top of the function, and also make use of FileSystem instead which will provide you with errors.

Reply:
Hope I've now cone this right